pkg/volume/azure_dd/azure_dd.go

```diff
@@ -0,0 +1,348 @@
+		return nil
+	}
+
+	globalPDPath := makeGlobalPDPath(b.plugin.host, b.diskName)
+
+	if err := os.MkdirAll(dir, 0750); err != nil {
+		glog.V(4).Infof("Could not create directory %s: %v", dir, err)
+		return err
+	}
+
+	// Perform a bind mount to the full path to allow duplicate mounts of the same PD.
+	options := []string{"bind"}
+	if b.readOnly {
+		options = append(options, "ro")
+	}
+	err = b.mounter.Mount(globalPDPath, dir, "", options)
+	if err != nil {
+		notMnt, mntErr := b.mounter.IsLikelyNotMountPoint(dir)
+		if mntErr != nil {
+			glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
+			return err
+		}
+		if !notMnt {
+			if mntErr = b.mounter.Unmount(dir); mntErr != nil {
+				glog.Errorf("Failed to unmount: %v", mntErr)
+				return err
+			}
+			notMnt, mntErr := b.mounter.IsLikelyNotMountPoint(dir)
+			if mntErr != nil {
+				glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
+				return err
+			}
+			if !notMnt {
+				// This is very odd, we don't expect it.  We'll try again next sync loop.
+				glog.Errorf("%s is still mounted, despite call to unmount().  Will try again next sync loop.", dir)
+				return err
+			}
+		}
+		os.Remove(dir)
+		return err
+	}
+
+	if !b.readOnly {
+		volume.SetVolumeOwnership(b, fsGroup)
+	}
+	glog.V(3).Infof("Azure disk volume %s mounted to %s", b.diskName, dir)
+	return nil
+}
+
+func makeGlobalPDPath(host volume.VolumeHost, volume string) string {
+	return path.Join(host.GetPluginDir(azureDataDiskPluginName), "mounts", volume)
+}
+
+func (azure *azureDisk) GetPath() string {
+	name := azureDataDiskPluginName
+	return azure.plugin.host.GetPodVolumeDir(azure.podUID, utilstrings.EscapeQualifiedNameForDisk(name), azure.volName)
+}
+
+type azureDiskUnmounter struct {
+	*azureDisk
+}
+
+var _ volume.Unmounter = &azureDiskUnmounter{}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (c *azureDiskUnmounter) TearDown() error {
+	return c.TearDownAt(c.GetPath())
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (c *azureDiskUnmounter) TearDownAt(dir string) error {
+	notMnt, err := c.mounter.IsLikelyNotMountPoint(dir)
+	if err != nil {
+		glog.Errorf("Error checking if mountpoint %s: %v", dir, err)
+		return err
+	}
+	if notMnt {
+		glog.V(2).Info("Not mountpoint, deleting")
+		return os.Remove(dir)
+	}
+
+	refs, err := mount.GetMountRefs(c.mounter, dir)
+	if err != nil {
+		glog.Errorf("Error getting mountrefs for %s: %v", dir, err)
+		return err
+	}
+	if len(refs) == 0 {
+		glog.Errorf("Did not find pod-mount for %s during tear down", dir)
+		return fmt.Errorf("%s is not mounted", dir)
+	}
+	c.diskName = path.Base(refs[0])
+	glog.V(4).Infof("Found volume %s mounted to %s", c.diskName, dir)
+
+	// lock the volume (and thus wait for any concurrrent SetUpAt to finish)
+	c.plugin.volumeLocks.LockKey(c.diskName)
```

Comment:
(ah, I see you call GetMountRefs again, this seems brittle and error-prone, just lock once, above the first GetMountRefs call and use that data here.)